File: src/main/java/org/kowboy/util/Vein.java
package org.kowboy.util;

import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.block.Block;
import org.bukkit.entity.Player;
import org.bukkit.util.BoundingBox;

import java.util.Optional;
import java.util.Set;

import static org.kowboy.util.BukkitUtils.formatLocation;

/**
 * A representation of a mineral vein.
 *
 * @author Craig McDaniel
 * @since 1.0
 */
public final class Vein implements Comparable<Vein> {
    // We will keep a +1 sized bounding box around the vein blocks.
    private BoundingBox box;

    // Approximate distance from the player. Only really used for sorting veins.
    private double distance;

    // How many blocks are in this vein.
    private int count;

    // All blocks in a vein are of the same type of material.
    private final Material type;

    // Note, we could keep a collection of all the blocks in the vein, but we don't have a use for that at this point.
    private Location location;

    public Vein(final Block block, final double distance) {
        this.distance = distance;
        location = block.getLocation();
        type = block.getType();
        box = BoundingBox.of(block).expand(1);
        count = 1;
    }

    public Vein(final Block block, final Player player) {
        this(block, player.getLocation().distance(block.getLocation()));
    }

    public int getCount() {
        return count;
    }

    public Location getLocation() {
        return location;
    }

    public Material getType() {
        return type;
    }

    /**
     * Does this vein contain the given block within it's bounding box? Is it the same material?
     *
     * @param block The block to test.
     * @return True if the block is part of the vein, false otherwise.
     */
    public boolean contains(final Block block) {
        return type.equals(block.getType()) && box.contains(BoundingBox.of(block));
    }

    /**
     * If this block belongs in the vein, add it.
     *
     * @param block The block to possibly add to this vein.
     * @return True if the block was added, false otherwise.
     */
    public boolean addBlock(final Block block) {
        if (!contains(block)) return false;
        count++;
        box = box.union(BoundingBox.of(block).expand(1));
        return true;
    }

    /**
     * Detects if two veins overlap.
     *
     * @param v The vein to test against this one.
     * @return True if the two veins overlap, false otherwise.
     */
    public boolean overlaps(Vein v) {
        // Keep in mind that the bounding box is a +1 block buffer around each vein. We therefore need to shrink
        // one of them by 1 or we risk a false positive (where blocks that are 1 block apart are considered overlapping).
        return this.box.overlaps(v.box.clone().expand(-1));
    }

    /**
     * Combines two veins together iff they overlap.
     *
     * @param v The vein to combine with this one.
     * @return The new combined vein iff they overlap, null otherwise.
     */
    public Vein combine(Vein v) {
        if (type.equals(v.type) && overlaps(v)) {
            this.count += v.count;
            this.distance = Math.min(this.distance, v.distance);
            this.box = this.box.union(v.box);
            return this;
        }
        return null;
    }

    /**
     * Finds a vein from the set that the given block should belong to and adds it to the vein.
     *
     * @param veins The set of veins to check.
     * @param block The block.
     * @return An Optional containing the vein that the given block was added to, otherwise the Optional will be empty.
     */
    public static final Optional<Vein> addToVein(final Set<Vein> veins, final Block block) {
        return veins.stream().filter(v -> v.addBlock(block)).findFirst();
    }

    public static final Set<Vein> addBlock(final Player player, final Set<Vein> veins, final Block block) {
        Optional<Vein> found = addToVein(veins, block);
        if (!found.isPresent()) {
            // Create a new vein and initialize it with this block.
            veins.add(new Vein(block, player.getLocation().distance(block.getLocation())));
        }
        return veins;
    }

    /**
     * Iterates through the second set of veins and for each vein, combines it with an existing vein in the first set if
     * there is an overlap, otherwise adds the vein to the first set.
     *
     * @param vs1 The first set of veins.
     * @param vs2 The second set of veins.
     * @return The combined set of veins.
     */
    public static final Set<Vein> combine(final Set<Vein> vs1, final Set<Vein> vs2) {
        for (Vein v2 : vs2) {
            Optional<Vein> found = vs1.stream()
                    .map(v1 -> v1.combine(v2))
                    .filter(v -> v != null)
                    .findFirst();
            if (!found.isPresent()) {
                // Add v2 to the first set
                vs1.add(v2);
            }
        }
        return vs1;
    }

    @Override
    public int compareTo(Vein v) {
        return Double.compare(this.distance, v.distance);
    }

    @Override
    public String toString() {
        // Example: diamond_ore * 4: (-13, 12, 72)
        return type.name().toLowerCase() + " * " + this.count + ": " + formatLocation(this.location);
    }
}

File: src/main/kotlin/org/kowboy/util/Vein.kt
package org.kowboy.util

import org.bukkit.ChatColor
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.block.Block
import org.bukkit.entity.Player
import org.bukkit.util.BoundingBox
import org.kowboy.util.ChatUtils.getColor
import java.util.*
import java.util.function.Predicate

/**
 * A representation of a mineral vein.
 *
 * @author Craig McDaniel
 * @since 1.0
 */
class Vein(block: Block,
    // Approximate distance from the player. Only really used for sorting veins.
    private var distance: Double) : Comparable<Vein> {

    // We will keep a +1 sized bounding box around the vein blocks.
    private var box: BoundingBox

    // How many blocks are in this vein.
    var count: Int
        private set

    // All blocks in a vein are of the same type of material.
    val type: Material = block.type

    // Note, we could keep a collection of all the blocks in the vein, but we don't have a use for that at this point.
    val location: Location = block.location

    constructor(block: Block, player: Player) : this(block, player.location.distance(block.location)) {}

    init {
        box = BoundingBox.of(block).expand(1.0)
        count = 1
    }

    /**
     * Does this vein contain the given block within it's bounding box? Is it the same material?
     *
     * @param block The block to test.
     * @return True if the block is part of the vein, false otherwise.
     */
    operator fun contains(block: Block): Boolean {
        return type == block.type && box.contains(BoundingBox.of(block))
    }

    /**
     * If this block belongs in the vein, add it.
     *
     * @param block The block to possibly add to this vein.
     * @return True if the block was added, false otherwise.
     */
    fun addBlock(block: Block): Boolean {
        if (!contains(block)) return false
        count++
        box = box.union(BoundingBox.of(block).expand(1.0))
        return true
    }

    /**
     * Detects if two veins overlap.
     *
     * @param v The vein to test against this one.
     * @return True if the two veins overlap, false otherwise.
     */
    fun overlaps(v: Vein): Boolean { // Keep in mind that the bounding box is a +1 block buffer around each vein. We therefore need to shrink
        // one of them by 1 or we risk a false positive (where blocks that are 1 block apart are considered overlapping).
        return box.overlaps(v.box.clone().expand(-1.0))
    }

    /**
     * Combines two veins together iff they overlap.
     *
     * @param v The vein to combine with this one.
     * @return The new combined vein iff they overlap, null otherwise.
     */
    fun combine(v: Vein): Vein? {
        if (type == v.type && overlaps(v)) {
            count += v.count
            distance = Math.min(distance, v.distance)
            box = box.union(v.box)
            return this
        }
        return null
    }

    override fun compareTo(v: Vein): Int {
        return distance.compareTo(v.distance)
    }

    override fun toString(): String { // Example: diamond_ore * 4: (-13, 12, 72)
        return getColor(type).toString() +
                type.name.toLowerCase() +
                ChatColor.GRAY +
                " * " +
                count +
                ": " +
                location.toBlockPoint()
    }
}

/**
 * Finds a vein from the set that the given block should belong to and adds it to the vein.
 *
 * @param block The block.
 * @return An Optional containing the vein that the given block was added to, otherwise the Optional will be empty.
 */
fun Set<Vein>.addToVein(block: Block): Vein? {
    return this.firstOrNull { vein -> vein.addBlock(block) }
}

/**
 * Adds a block to an existing vein if it can (if it belongs to one), or else creates a new vein and adds it
 * to the set.
 *
 * @param block The block to add.
 * @param player The player - used to calculate distance to the block.
 * @return The updated set of veins.
 */
fun Set<Vein>.addBlock(block: Block, player: Player): Set<Vein> {
    if (null == addToVein(block)) {
        return this + Vein(block, player.location.distance(block.location))
    }
    return this
}

/**
 * Iterates through the second set of veins and for each vein, combines it with an existing vein in the first set if
 * there is an overlap, otherwise adds the vein to the first set.
 *
 * @param vs2 The second set of veins.
 * @return The combined set of veins.
 */
fun Set<Vein>.combine(vs2: Set<Vein>): Set<Vein> {
    return vs2.fold(this) { acc, v2 ->
        val found = acc.asSequence().mapNotNull { v1 -> v1.combine(v2) }
                .firstOrNull()
        if (null == found) {
            acc + v2
        } else {
            acc
        }
    }
}